Include TypeScript definitions in bundle

rexxars opened the pull request:
Treat this as a work-in-progress that we can use to base our discussions from.
Addresses #3369

In essence, this lifts the types from the `@types/sharp` module into the official package, along with the tests that were in that package as well.

Because the types reference node.js types such as `Buffer`, we also need to add a new dependency: `@types/node`. The DefinitelyTyped modules uses the wildcard version range (`*`) in order to prevent multiple copies of the package in the module tree, which often leads to hard to diagnose problems. An alternative would be for us to use `>=14.0.0` or similar. Open to suggestions.

As @lovell pointed out in #3369, this should be considered a minor breaking change, so should likely be released as 0.32.0. When that happens, we can also deprecate `@types/sharp` using the [official process](https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package).

Some thoughts:
- While I don't _know_ of any errors in the typings, there is always a chance that there are. I wouldn't be surprised to see a few more PRs popping up to address issues.
- There is starting to be a bit of duplication in documenting things - there is a chance that the jsdoc and the tsdoc is slightly out of sync. Ideally we'd find a better process here, or even consider doing "typescript first" and compiling to js (this is probably controversial - just thinking aloud).
- The "type tests" are somewhat spotty in coverage, and very unstructured. Unless there are plans to migrate to TypeScript fully, I would probably want to restructure these tests a bit over time - splitting them into logical, "sorted" sections of some kind. I'm not sure if it's worth waiting for this work to be completed before a release however - open to suggestions.

lovell commented on package.json:
```diff
@@ -130,6 +132,7 @@
     "vips"
   ],
   "dependencies": {
+    "@types/node": "*",
```
Do we need this? My understanding is that it's the responsibility of a consuming TypeScript-based app to define this dependency, rather than doing so in every package that happens to provide type definitions. This approach also avoids the versioning problem you mention.
```sh
$ du -sh node_modules/@types/node
1.9M	node_modules/@types/node
```

rexxars replied:
1.9M worth of typings 👀  Color me surprised.

I honestly don't quite know what the "right" approach is here, or if there even is one:
- A direct dependency will work and have typescript be happy when using the module
- A peer dependency marked as optional might be one option if we want to be explicit about what it depends on
- Dropping it altogether (or rather, moving it to a dev dependency - I think we'll need it for tsd to work) is probably also acceptable, since as you say - people using sharp + typescript is likely to install `@types/node` in their app anyway

lovell replied:
Let's go with including it in `devDependencies`, which I think is the least intrusive option. We should also add a section to the installation docs about TypeScript support, including the need to install `@types/node` yourself. I can add this later.

I'm also bracing for future PRs that attempt to move `@types/node` to `dependencies` :)

rexxars replied:
> I'm also bracing for future PRs that attempt to move `@types/node` to `dependencies` :)

Haha, it is bound to happen! 👍

rexxars replied:
Moved to dev dependencies and force-pushed.

lovell reviewed (commented):
Thank you for making a start on this, as you suggest it will need to be part of v0.32.0. I've left one question inline.

rexxars commented:
Let me know if you want any more changes done to this!

rexxars commented:
I have rebased this PR and lifted the changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63711 - which incorporates the changes that were added in sharp v0.31.3, along with the added tests. Squashed the changes into the previous commits to keep the commits clean. Also added @akwiatek as an author of the typings.

rexxars commented:
Nothing else I want to add! Rebased on `flow` and fixed the conflict 👍

lovell commented:
Thank you very much Espen, these will be published as part of v0.32.0.

Thanking you also in advance should I need to draw on your expertise to help review any PRs that arise due to this :+1:

lovell commented:
The definition "diff" for the forthcoming v0.32.0 release is in commit https://github.com/lovell/sharp/commit/a532659b0fefbeb0c2c18eba7f56dd36932abaa4 - if you spot anything else out of sync, please do open a PR.
